services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/UploadWithKnownContentLengthHelper.java

```diff
@@ -90,98 +95,173 @@ public CompletableFuture<PutObjectResponse> uploadObject(PutObjectRequest putObj
+
+        if (resumeToken == null) {
+            partSize = genericMultipartHelper.calculateOptimalPartSizeFor(contentLength, partSizeInBytes);
+            partCount = genericMultipartHelper.determinePartCount(contentLength, partSize);
+            if (partSize > partSizeInBytes) {
+                log.debug(() -> String.format("Configured partSize is %d, but using %d to prevent reaching maximum number of "
+                                              + "parts allowed", partSizeInBytes, partSize));
+            }
+
+            log.debug(() -> String.format("Starting multipart upload with partCount: %d, optimalPartSize: %d", partCount,
+                                          partSize));
+        } else {
+            partSize = resumeToken.partSize();
+            partCount = (int) resumeToken.totalNumParts();
+            long numCompleted = resumeToken.numPartsCompleted();
 
-        long optimalPartSize = genericMultipartHelper.calculateOptimalPartSizeFor(contentLength, partSizeInBytes);
-        int partCount = genericMultipartHelper.determinePartCount(contentLength, optimalPartSize);
-        if (optimalPartSize > partSizeInBytes) {
-            log.debug(() -> String.format("Configured partSize is %d, but using %d to prevent reaching maximum number of parts "
-                                          + "allowed", partSizeInBytes, optimalPartSize));
+            log.debug(() -> String.format("Resuming multipart upload with completedPartCount: %d, remainingPartCount: %d, "
+                                          + "partSize: %d, ", numCompleted, partCount - numCompleted , partSize));
         }
 
-        log.debug(() -> String.format("Starting multipart upload with partCount: %d, optimalPartSize: %d", partCount,
-                                      optimalPartSize));
+        MpuRequestContext mpuRequestContext =
+            new MpuRequestContext(request, contentLength, partSize, uploadId, existingParts);
 
-        MpuRequestContext mpuRequestContext = new MpuRequestContext(request, contentLength, optimalPartSize, uploadId);
+        KnownContentLengthAsyncRequestBodySubscriber subscriber =
+            new KnownContentLengthAsyncRequestBodySubscriber(mpuRequestContext, returnFuture);
+
+        attachSubscriberToObservable(subscriber, request.left());
 
         request.right()
                .split(b -> b.chunkSizeInBytes(mpuRequestContext.partSize)
                             .bufferSizeInBytes(maxMemoryUsageInBytes))
-               .subscribe(new KnownContentLengthAsyncRequestBodySubscriber(mpuRequestContext,
-                                                                           returnFuture));
+               .subscribe(subscriber);
+    }
+
+    private HashMap<Integer, CompletedPart> identifyExistingPartsForResume(String uploadId, PutObjectRequest putObjectRequest) {
+        HashMap<Integer, CompletedPart> existingParts = new HashMap<>();
+        int partNumberMarker = 0;
+
+        while (true) {
```

Comment:
Can we use `listPartsPaginator` instead?

Reply:
updated to use `listPartsPaginator`